Add ZeroMQ notifications

promag opened the pull request:
Continues Johnathan Corgan and Jeff Garzik's work. Supercedes #4594 and #5303.
As discussed in #6072 and #5328.

Commit:
Add ZeroMQ support. Notify blocks and transactions via ZeroMQ

Continues Johnathan Corgan's work.
Publishing multipart messages

Bugfix: Add missing zmq header includes

Bugfix: Adjust build system to link ZeroMQ code for Qt binaries

ajweiss commented on src/zmqports.cpp:
```diff
@@ -0,0 +1,175 @@
+
+static void zmqError(const char *str)
+{
+  LogPrint("ZMQ error: %s, errno=%s\n", str, zmq_strerror(errno));
```
This entire file needs to be retabbed.

jonasschnelli replied:
L39: s/`LogPrint`/`LogPrintf` otherwise you start introducing a debug category with string identifier `"ZMQ error: %s, errno=%s\n"`.

promag replied:
Fixed in upcoming commit

ajweiss commented:
Are you planning on supplying an RPC/regression test for this?

promag commented:
That work was in previous PR. I can't see a use case where this is useful.

sipa commented on src/ui_interface.h:
```diff
@@ -95,6 +97,12 @@ class CClientUIInterface
     boost::signals2::signal<void (const uint256& hash)> NotifyBlockTip;
+
+    /** New block has been accepted */
+    boost::signals2::signal<void (const CBlock& block)> NotifyAcceptBlock;
```
As mentioned in one of the previous PRs, block acceptance is really not interesting or useful. It does not mean the block is part of the best chain, or even that it is valid. Use the NotifyBlockTip signal place, which indicates a new tip of the best chain was accepted.

promag replied:
Fixed in upcoming commit

ajweiss commented:
As far as I know, no RPC tests for ZMQ support have yet been PR'd.

I think there might be some confusion here.  When I say RPC tests, I mean the Python regression tests that live qa/rpc-tests tree.  They're pretty much only called RPC tests because of history and the primary way they interface with bitcoind.  They exercise and verify many large bits of functionality and are always useful!  Case in point, when REST support was originally added, it had some pretty serious bugs that weren't caught because there was no RPC test.  See PR #5379 for an example of how RPC tests for the REST interface were implemented.

I suppose RPC tests could be PR'd separately, but I bet this will go in faster and with more enthusiasm if tests are added.

jonasschnelli commented:
@promag Nice that you have started recycle/update/rebase this!
I have plans to test this during this week and therefore would be willing to write some RPC'like tests.

jonasschnelli commented on doc/zmq.md:
```diff
@@ -0,0 +1,145 @@
+Currently, the ZeroMQ facility only needs to have the ZeroMQ endpoint
+specified:
+
+    $ bitcoind -zmqpub=tcp://127.0.0.1/28332
```
this should be: `bitcoind -zmqpub=tcp://127.0.0.1:28332` (dash instead slash before port)

promag replied:
fixed in upcoming commit

jonasschnelli commented on src/zmqports.cpp:
```diff
@@ -0,0 +1,175 @@
+  }
+
+  uiInterface.NotifyAcceptBlock.connect(ZMQPublishBlock);
+  uiInterface.NotifyRelayTx.connect(ZMQPublishTransaction);
```
Does it make sense to use the uiInterface `CClientUIInterface` as signaling layer for zmq? Or would it be better to introduce a custom signaling class/struct?

promag replied:
I believe this discussion should happen in a different PR.

jonasschnelli commented on src/rpcserver.cpp:
```diff
@@ -282,6 +282,9 @@ static const CRPCCommand vRPCCommands[] =
     { "network",            "ping",                   &ping,                   true  },
 
+    /* ZMQ notification */
+    { "zmq",                "getzmqurl",              &getzmqurl,              true  },
```
What's the use case of `getzmqurl`? Why should someone need to get the ZMQ URL over RPC?

promag replied:
removed in upcoming commit

jonasschnelli replied:
I didn't say it has to be removed. I just don't see a usecase for this. Maybe @jgarzik can roll this up?

jonasschnelli replied:
Or maybe something for @jmcorgan

jonasschnelli commented on src/zmqports.cpp:
```diff
@@ -0,0 +1,175 @@
+
+  // Serialize block
+  CDataStream ss(SER_NETWORK, PROTOCOL_VERSION);
+  ss.reserve(1000000); // FIXME use defined constant
```
I think this line can be removed. Also L112.
See also: https://github.com/bitcoin/bitcoin/blob/master/src/rpcblockchain.cpp#L308 and https://github.com/bitcoin/bitcoin/blob/master/src/rest.cpp#L182

promag replied:
I've changed this to use zeromq multipart messages. Changed in upcoming commit

promag commented on src/ui_interface.h:
```diff
@@ -95,6 +97,9 @@ class CClientUIInterface
     boost::signals2::signal<void (const uint256& hash)> NotifyBlockTip;
+
+    /** New transaction to relay */
+    boost::signals2::signal<void (const CTransaction& tx)> NotifyRelayTx;
```
I don't think NotifyRelayTx should exist at all.
As suggested by @jonasschnelli, instead of using CClientUIInterface signals we could use CValidationInterface SetBestChain and SyncTransaction

promag commented on src/zmqports.cpp:
```diff
@@ -0,0 +1,183 @@
+        LOCK(cs_main);
+    
+        CBlockIndex* pblockindex = mapBlockIndex[hash];
+        if(!ReadBlockFromDisk(blk, pblockindex))
```
I believe this is the right way of getting a block.

promag commented:
I'm planning to add the option -zmqformat=raw|json. This way a subscriber doesn't need to rely on 3rd libs to process the messages. What do you think?

jonasschnelli committed:
QA: Add ZeroMQ RPC test

jonasschnelli commented on src/zmqports.h:
```diff
@@ -0,0 +1,26 @@
+// Global state
+extern bool fZMQPub;
+
+#if ENABLE_ZMQ
```
Is the reason why the `#ifdefs` are here instead of wrapping the code-part in `init.cpp` to lower the #ifdef in `init.cpp`? Maybe also something for @jgarzik to answer.

promag replied:
FYI https://github.com/bitcoin/bitcoin/pull/4594#issuecomment-50352710

jonasschnelli replied:
@promag: Thanks. Yes this makes sense.

jonasschnelli commented:
Would be nice if you could cherry-pick / pull-in this RPC/ZMQ test case:
https://github.com/jonasschnelli/bitcoin/commit/91334e70ee5b2713ebbb7d5dbd6a5a29936aba66

jonasschnelli commented on src/init.cpp:
```diff
@@ -1022,6 +1029,11 @@ bool AppInit2(boost::thread_group& threadGroup)
 
+    if (mapArgs.count("-zmqpub"))
+    {
+        ZMQInitialize(mapArgs["zmqpub"]);
```
typo: should be `ZMQInitialize(mapArgs["-zmqpub"]);` (mind the dash)

jonasschnelli commented:
@promag: I would not add JSON support and keep ZMQ as clean and low-level as possible. Keeping it bin only can make things faster and with JSON there is always a risks of things get handled different during enc-/decoding of JSON streams depending on the library you use.

jonasschnelli commented:
Tested reviewed leak-checked ACK.
nits: usage of `uiInterface` as "signal layer"

jonasschnelli commented on src/zmqports.cpp:
```diff
@@ -0,0 +1,164 @@
+    CBlock blk;
+    {
+        LOCK(cs_main);
+    
```
nit: L128 has whitespace.

jonasschnelli commented:
nit:
signaling disconnect is missing

```diff
diff --git a/src/zmqports.cpp b/src/zmqports.cpp
index d083292..bd1229c 100644
--- a/src/zmqports.cpp
+++ b/src/zmqports.cpp
@@ -155,6 +155,9 @@ void ZMQShutdown()
 
         zmq_ctx_destroy(zmqContext);
         zmqContext = 0;
+
+        uiInterface.NotifyBlockTip.disconnect(ZMQPublishBlock);
+        uiInterface.NotifyRelayTx.disconnect(ZMQPublishTransaction);
     }
 
     fZMQPub = false;
```

promag commented:
@jonasschnelli Regarding the output formats, I don't think having JSON is that bad, it is just an option.
Actually I think we could support 3 formats for now:
 - network: currently supported
 - json: needs BlockToUnit to convert to JSON (similar to RPC getblock)
 - hash: just send the block hash or txid
The last one allows a consumer to queue the notification to process it later.

jonasschnelli commented:
@promag I agree with 1 (p2p) and 3 (hash). IMO the ZMQ interface is a push/notify interface (for now). If someone like to get JSON output he could do a RPC request with received data. Adding JSON format over ZMQ level would just encourage user "to go the wrong way".
Adding a way of only receiving hashes over ZMQ would be good.
But i'm ready to hear other opinions.

sipa commented:
Specifically for blocks, I'm not sure that pushing full blocks is even useful, unless we make it push all new blocks (including intermediary reorged ones). The interesting information is "there is a new block chain tip, and this is it".

promag commented:
@sipa consider the following use case: a subscriber wants to know if there are transactions (in the new block chain tip) with outputs targeting a set of addresses.

promag commented:
I'm not sure if publishing a block of 20MB is problematic.

sipa commented:
That use case would not be covered, as the current code only reports the
new tip, not potential intermediate blocks that were connected as part of a
reorganization.

A possibility is reporting all new chain blocks, but what if this (due to a
fork bug, perhaps) ends up reporting 100s of megabytes?

promag commented:
@sipa right, so publishing the new tip hash and having the consumer walk the block chain would be more appropriate/secure.

jonasschnelli commented:
For now i see the ZMQ extension as a notification-only one-way interface. I guess it's more optimized if a user gets informed about a new block **hash** (or new mempool tx **hash**) and then he decides if he likes to load the block with or without tx details over REST or RPC.

promag commented:
Agree on network binary format and hash format? Other publish strategies can be discussed later after this PR.

jonasschnelli commented:
@promag: i think you need to get rid of commit 454285518915d73e55674cf702bd0372a2da266e

promag commented:
Should we use CValidationInterface instead of CClientUIInterface?

sipa commented:
Yes.

promag commented:
Agree on CZMQPublisher : public CValidationInterface ?

promag commented:
Pending questions:
 1. support network binary format and hash format?
 2. class name CZMQPublisher : public CValidationInterface?
 3. use CValidationInterface::SetBestChain instead of CClientUIInterface::NotifyBlockTip?

sipa commented:
1. Hash is enough for me. Binary makes sense if we make sure the full set of newly accepted blocks is announced.

2. Sounds good.

3. SetBestChain won't work, as it is only called after the best chain on disk is changed (which is infrequently). I'll write a patch to add the signals you need to CValidationInterface.

promag commented:
@sipa At the moment CValidationInterface is very related to CWallet, but it shouldn't right?

sipa commented:
@promag It used to be called CWalletInterface. We renamed it because it shouldn't be used exclusively for wallets.

promag commented on src/zmqports.h:
```diff
@@ -0,0 +1,39 @@
+
+protected: // CValidationInterface
+    void SyncTransaction(const CTransaction &tx, const CBlock *pblock);
+    void UpdatedBlockTip(const CBlock &block);
```
waiting @sipa patch, for now assuming UpdatedBlockTip exists in CValidationInterface.

promag commented on src/zmqports.cpp:
```diff
@@ -0,0 +1,195 @@
+        // Send multipart message
+        int rc = Publish("BLK", 3, ZMQ_SNDMORE);
+        if (rc==0)
+            Publish(hash.begin(), hash.size(), 0);
```
@jonasschnelli is this OK?

jonasschnelli replied:
LGTM.

promag commented:
Later will address https://github.com/bitcoin/bitcoin/pull/5303#discussion_r26264945

jonasschnelli commented:
Will do additional tests when `UpdatedBlockTip` is ready.

promag commented:
It seems Travis is not building with zeromq. Can this be addressed?

theuni committed:
Depends: Add ZeroMQ package

theuni commented:
@promag Yes. You'll need to add the dependency. You can pull this commit: https://github.com/theuni/bitcoin/commit/3661302abfca852a871b8b3cb406d5d10dfc33ae to get travis to test.

Note, however, that the windows build currently fails because zeromq refuses to static-link with mingw. I chose to leave this broken rather than disabling the windows build because we'll have to come up with a plan, even if it's just to explicitly disable it there.

promag commented:
@sipa if you like I can add the new signal to CValidationInterface. I suppose I can emit it before the NotifyBlockTip.

promag commented:
@theuni BTW why static linking? and why not dynamic linking on travis because of http://docs.travis-ci.com/user/ci-environment/#Messaging-Technology

theuni commented:
@promag Travis builds exactly as we build for release. The current build failure means that we wouldn't be able to package zeromq into a windows binary. Until that's fixed, failing is the right thing to do imo. The question is: is it worth it to try to work around their mingw situation, or just disable it for windows.

jonasschnelli commented:
Having ZMQ for windows as well would be nice in near future. With ZMQ we could place a (daemon)script into the /contrib folder and deprecate the -notify features. IIRC there where serval issues with the -notify commands on windows.

@promag: I think it would be nice if you could complete this and try to add the new signal to `CValidationInterface()`

Commit:
Add UpdatedBlockTip signal to CMainSignals and CValidationInterface

promag commented:
@theuni should we disable zmq in windows?
ping @sipa

jonasschnelli commented on src/init.cpp:
```diff
@@ -345,6 +357,10 @@ std::string HelpMessage(HelpMessageMode mode)
 
+    strUsage += HelpMessageGroup(_("ZeroMQ publisher options"));
+    strUsage += HelpMessageOpt("-zmqpub=<endpoint>",_("Publish blocks and transactions on 'endpoint'"));
+    strUsage += HelpMessageOpt("-zmqformat=hash|network>",_("Publish format is either hash (default) or network"));
```
s/-zmqformat=hash|network**>**/-zmqformat=**<**hash|network**>**/

promag commented:
Disabled zeromq in windows because zmq fails to build as a static lib in mingw.

jtimon commented:
utACK

luke-jr commented:
Why not publish both formats and let clients choose which to subscribe to?

theuni commented:
@promag That only disables the windows build for travis, not for anyone actually building for windows.

Please instead use something like:
```
packages:=boost openssl
packages_darwin:=zeromq
packages_linux:=zeromq
```

Please also document the fact that it's unusable for Windows due to upstream build issues.

promag commented:
@luke-jr you mean one address per format?
@theuni right, will do.

luke-jr commented:
@promag I mean using ZeroMQ's topics so clients can filter which they want.

promag commented:
@luke-jr Actually I was thinking removing the option to publish the entire transaction/block. The block will have a maximum size of 20mb. And writing all formats to the socket sequentially can cause some lag each time there is a new block. Also, at the moment, writing to the socket happens in the main thread, and the publisher should be as fast as possible.

At the moment we use the hash format, and the hash is enqueued in a persistent/reliable queue so that it can be processed later on a easy pace. Then the RPC interface is used to get more details about it. We think the load of bitcoind is much lower than publishing the entire data.

jtimon commented:
In fact, apart from several formats we could allow the option of using either PUB or PUSH instead of forcing it to be PUB?
PUSH should be much faster than PUB, both with 1 MB and 20 MB blocks (which is something completely orthogonal to this PR).

promag commented:
@jtimon IMHO this notification mechanism should be as simple as possible. Just publishing the event/hash is enough for the downstream to make RPC calls to get more information about blocks/transactions, including computed data which is not in the raw data, for instance, confirmations etc.

In our use case we tried to avoid RPC calls upstream, but ended giving up because we didn't want to miss the blockchain maintenance of bitcoind, ie, handling forks, double spends, etc. So, with this notifications we just avoid notify-* scripts and RPC polling.

Moreover, I would never send the raw data, because it is too much and requires the downstream to "know" how to parse the raw data. Handling JSON RPC is much easier.

Regarding the socket type, I wouldn't choose a socket where the send operation is blocking as this could lead to, either internal queue in a separate thread, or making bitcoind malfunction. Therefore, IIRC PUB/SUB is the only type that has non-blocking send.

jtimon commented:
> @jtimon IMHO this notification mechanism should be as simple as possible. 

IMO it should be as flexible as possible. That's what I love most about zmq, anyway.

> In our use case we tried to avoid RPC calls upstream, but ended giving up because we didn't want to miss the blockchain maintenance of bitcoind, ie, handling forks, double spends, etc. So, with this notifications we just avoid notify-* scripts and RPC polling.

That's seems good for your use case. Other uses of zmq could be to precisely replace rpc calls (although we're not supporting REQ/REP in this PR and that's fine). Remember that RPCs also take locks and all that. Anyway, please don't assume that everybody wants to use the zmq integration in the same way you do.

> Moreover, I would never send the raw data, because it is too much and requires the downstream to "know" how to parse the raw data. Handling JSON RPC is much easier.

Zmq also supports json, but I can see some potential consumers preferring raw data.

> Regarding the socket type, I wouldn't choose a socket where the send operation is blocking as this could lead to, either internal queue in a separate thread, or making bitcoind malfunction. Therefore, IIRC PUB/SUB is the only type that has non-blocking send.

I didn't know that PULL/PUSH was blocking, but doesn't make much sense to me. Then why wouldn't people just use a PUB/SUB with a single subscriber that then pushes to another sockets (to avoid making the original send blocking)?
Can you link me to the zmq documentation that says PUSH sends are blocking but PUB are not?
Even if that's the case, it may be worth it for some use cases.

Anyway, your arguments for not supporting topics in PUB/SUB (as Luke-Jr proposes) seem invalid to me, specially if you were assuming that only hashes will be sent.

promag commented:
@jtimon From http://api.zeromq.org/4-0:zmq-socket

> When a ZMQ_PUSH socket enters the mute state due to having reached the high water mark for all downstream nodes, or if there are no downstream nodes at all, then any zmq_send(3) operations on the socket shall block until the mute state ends or at least one downstream node becomes available for sending; messages are not discarded.

promag commented:
IMHO the true purpose of this PR is to eventually avoid/deprecate notify scripts. It should provide the CValidationInterface signals to the outside world. I agree with all that flexibility/options but for me it is out of scope of this PR.

jtimon commented:
Oh I see, of course, push is blocking when you don't have ready pullers...that has a simple solution: have pullers. That's a problem for the user to solve. IMO not a good reason not to support PUSH ( which is so similar to PUB in all other things).

About the scope of this PR, sure, this doesn't have to do everything, but for example supporting PUB/SUB topics definitely seems within scope.

If we're going to leave pushers for later, that's fine, sunce we want a different option to activate it anyway. Still, the current options could indicate that this is for PUB explicitly, for the options to be more clear when/if we add PUSH (or even REP) later. Also, one should be able to configure the transactions and  locks events independently.

jtimon commented:
So in summary, I would separate zmqpub into zmqpubtx and zmqpubblock or something similar, and I would add support for topics, everything else can be left for later.

promag commented:
What you mean by zmqpubtx and zmqpubblock? Two pub sockets binded to different addresses?

jmcorgan commented:
On Fri, May 29, 2015 at 2:02 AM, João Barbosa <notifications@github.com>
wrote:

> Regarding the socket type, I wouldn't choose a socket where the send
> operation is blocking as this could lead to, either internal queue in a
> separate thread, or making bitcoind malfunction. Therefore, IIRC PUB/SUB is
> the only type that has non-blocking send.
>

This is precisely the reason PUB was chosen when I originally implemented
this.

-- 
Johnathan Corgan
Corgan Labs - SDR Training and Development Services
Intro to SDR Classes - June 4-5, Columbia MD, June 29-30, El Segundo, CA
http://corganlabs.com

ajweiss commented:
This pull already implements topics via construction of multipart messages.  They are "BLK" and "TXN".  If the zmq library on the publisher side is recent enough, it will filter on the publisher side as per the subscriber's topic preferences.  

This, however, doesn't help all that much as ideally you'd want something that can tell whether or not anyone is subscribed to a topic so you can save the cost of serialization when it's not necessary.  AFAIK, there's no simple way to ask zmq if anyone is subscribed or what topics are subscribed.

Whether or not the cost of serialization/zmq_sending on all transactions and blocks is enough to justify adding configuration options (and translation/documentation overhead) to selectively turn block or transaction publishing on/off is an open question.  I have not measured it.

jtimon commented:
> What you mean by zmqpubtx and zmqpubblock? Two pub sockets binded to different addresses?

Yes. well in fact maybe you want to connect both to the same socket, but yes the goal would be that you are able to configure them independently (maybe you confiure blocks but not transactions or viceversa).

luke-jr commented:
If the goal is to merely notify of hashes, I suggest an external program triggered by -*notify may be better.

promag commented:
@luke-jr notify scripts only support wallet transactions and blocks. There is no support for mempool notifications (#6072).

Beside that, executing a script for each hash is overkill, specially when lots of transactions change state like when the blockchain is incremented. Also, notify scripts execute in the same machine the bitcoind runs whereas sockets provide much more flexibility.

jtimon commented:
So does it make sense to be able to configure independent sockets for transactions and blocks?
You could still connect both to the same port instead of binding them, presumably used in combination of a zmq Forwarder or binding a single subscriber.

luke-jr commented:
@promag Ok, you've made your point. I suggest at least offering RPC over ZMQ though, but that's I guess out of scope here.

jmcorgan commented:
RPC over ZMQ is indeed out of scope on this pull request, but definitely worth considering.

promag commented:
Agree on discarding RPC for now.

So I think we could allow arguments -zmq{type}{format}{object}={address} where
 - type = pub (push in the future)
 - format = hash | raw
 - object = transaction | block
 - address = a valid zmq socket address

jtimon commented:
> Agree on discarding RPC for now.
>
> So I think we could allow arguments -zmq{type}{format}{object}={address} where
>
> type = pub (push in the future)
> format = hash | raw
> object = transaction | block
> address = a valid zmq socket address

Sounds great to me.

promag commented:
Ops, forgot to add the topic option, which is important if an address is shared. Unless if topic the the same has the object.

promag commented:
@jonasschnelli @jgarzik please review the refactor, still wip

jonasschnelli commented on src/init.cpp:
```diff
@@ -349,9 +361,12 @@ std::string HelpMessage(HelpMessageMode mode)
 #endif
 
+    strUsage += HelpMessageGroup(_("ZeroMQ publisher options"));
+    strUsage += HelpMessageOpt("-zmq<type><format><object>=<address>",_("Enable notification through zmq socket <address> (type=pub, format=hash|raw, object=block|transaction)"));
```
It looks like that this is not implemented like that. Example: `zmqpubrawblock` is not supported. Only `zmqpubhashblock` and `zmqpubhashtransaction` is supported IMO. At least the should be a warning or error if you try to add a zmq channel which is not supported.

promag replied:
As stated it's just an initial version to ask for comments. The usage help will havê the supported options.

jonasschnelli commented:
**This PR needs some love.**

1. documentation out of date (or lets say out of sync because it's not yet available)
2. `./contrib/zmq/zmq_sub.py` out of date, doesn't run.
3. `./contrib/zmq/zmq_sub.py` as no `+x` (executable) flag like `linearize.py`, etc.
4. zmq rpc test does not run, out of date
5. the factory thing as well as the flexible `-zmq<pub><...>` feels over the top and should throw an error if given "format" could not generate a listener.

I really think ZMQ within bitcoin-core would be a great step, but this PR lacks here and there at the moment.

promag commented:
Agree but before polishing and squash I want to be sure that the refactor is acceptable.

Instead of having lots of switch/ifs I added a notifier type for each type of notification. Later other types of notifications can be added easilly. If it's OK then I'll update everything.

jonasschnelli commented:
Another thing:
If i add `-zmqpubhashblock=tcp://127.0.0.1:28332 -zmqpubhashtransaction=tcp://127.0.0.1:2833` (both notification types) it won't send blockhashes through zmq (didn't analyze in detail).

Conceptual:
I'm not sure if the flexible cmd arg is the right thing. It is like settings config values within a config arg key which feels somehow strange.
Also unsure why lots of things are compiled in when `--enable-zmq=no` (cmd args get parsed and factory gets created even when having `--enable-zmq=no` during compile time. This feels wrong).

Feel free to pull in the zmq test / zmq example tool fixes:
https://github.com/jonasschnelli/bitcoin/commit/fe2e286856eb35e0614fa5e1da757ff44ba7eeb1

promag commented:
@jonasschnelli Flexible command args is just a way to explain it here because of the multiple combinations. At then end the usage help will have listed just the notifiers supported. So, for instance, if command argument `-zmqpubhashblock` is defined then it will enable publishing hash blocks.

I'll follow https://github.com/bitcoin/bitcoin/pull/4594#issuecomment-50356713 to avoid compiling zmq notifications when `--enable-zmq=no`.

I'll rebase your commit and wrap everything.

jonasschnelli commented:
Needs rebase.

promag commented:
@jonasschnelli done

wozz commented on contrib/zmq/zmq_sub.py:
```diff
@@ -0,0 +1,38 @@
+import zmq
+
+port = 28332
+port = 4849
```
extra port

wozz commented on doc/zmq.md:
```diff
@@ -0,0 +1,158 @@
+        zmqContext.destroy()
+
+
+This example is provided in the contrib/zmq directory of the source
```
this is out of sync

promag commented:
@jonasschnelli @wozz done

nunofgs commented:
@jtimon, @sipa, @jonasschnelli: is there a timeframe for merging this? Would love to see this in 0.12!

rebroad commented:
I don't see that anyone has asked this, and it doesn't seem to be mentioned, but: who would use this? Is it a majority or a small minority that would? If the latter, can this functionality be added with some sort of 3rd party utility instead? If so, then I don't think it ought to be included in Bitcoin Core as it would add bloat and reduce security.

jonasschnelli commented:
@rebroad: I think consensus is that ZMQ (0mq) is desirable within bitcoin core (https://github.com/bitcoin/bitcoin/pull/2415, https://github.com/bitcoin/bitcoin/pull/5303) . At the moment, bitcoin-core users (including merchants, etc.) need to use `-blocknotify` and co. to get informed when a new block is available.

ZMQ would potential also allow to separate the wallet from the core and let them run in different processes.

promag commented:
@rebroad If you follow the discussion, and related pull requests, you will find enough arguments to deprecate notify scripts and have a non-intrusive notification interface.

jtimon commented:
@nunofgs I don't know, whenever is merged? I already ut ACKedm I had some nits that go resolved... re ut ACK.

jonasschnelli commented on doc/zmq.md:
```diff
@@ -0,0 +1,155 @@
+
+The ZeroMQ facility implements a notification interface through a
+set of specific notifiers. Currently there are notifiers that publish
+blocks and transactions. The  supports a set of no binds a "publish" port that broadcasts all newly
```
Linebreak is wrong.

D9B4BEF9 commented:
I'm struggling to get this working. Running as per the example in `zmq.md`:

    ./src/bitcoind -printtoconsole -zmqpubhashtx=tcp://127.0.0.1:28332

The node doesn't emit any notifications, the `msghand` thread spins endlessly on `CZMQAbstractNotifier::NotifyBlock()` and as a result a clean shutdown can never be completed. I don't have the time to debug it just now, just reporting that it doesn't seem to be completely stable on all platforms.

jonasschnelli commented on doc/zmq.md:
```diff
@@ -0,0 +1,155 @@
+doing so will result in no messages arriving.
+
+Here is a small example, in the Python language, using the
+*python-zmq* wrapper:
```
IMO better mention `contrib/zmq/zmq_sub.py` here. Otherwise we have to maintain two python examples.

promag replied:
@jonasschnelli I had the same feeling, will do it.

jonasschnelli commented:
Hmm.. can't compile this PR.

OSX:
```
  "CZMQNotificationInterface::Initialize()", referenced from:
      AppInit2(boost::thread_group&, CScheduler&) in libbitcoin_server.a(libbitcoin_server_a-init.o)
  "CZMQNotificationInterface::CreateWithArguments(std::__1::map<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::less<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > > const&)", referenced from:
      AppInit2(boost::thread_group&, CScheduler&) in libbitcoin_server.a(libbitcoin_server_a-init.o)
  "CZMQNotificationInterface::Shutdown()", referenced from:
      Shutdown() in libbitcoin_server.a(libbitcoin_server_a-init.o)
```

linux:
```
zmq/zmqpublishnotifier.cpp: In function ‘int zmq_send_multipart(void*, const void*, size_t, ...)’:
zmq/zmqpublishnotifier.cpp:30:61: error: ‘zmq_msg_send’ was not declared in this scope
zmq/zmqnotificationinterface.cpp: In member function ‘void CZMQNotificationInterface::Shutdown()’:
zmq/zmqnotificationinterface.cpp:119:33: error: ‘zmq_ctx_destroy’ was not declared in this scope
```

promag commented:
@D9B4BEF9 what's the platform and zmq version?

D9B4BEF9 commented:
@promag OSX, ZMQ 4.0.5.

jonasschnelli commented:
@promag: 
linux: libzmq1 2.2.0+dfsg-2 (debian)
osx: zeromq/4.0.5_2

luke-jr commented:
zmqnotificationinterface.cpp and zmqpublishnotifier.cpp are missing #include <zmq.h>

zmqpublishnotifier.cpp is missing copyright headers

D9B4BEF9 commented:
Alright, some quick progress into what but not why.  

`./src/bitcoind -printtoconsole -zmqpubhashblock=tcp://127.0.0.1:28332'`

The node comes up, those two sockets listen as expected, and then the node catches up on the latest block. The worker gets stuck emitting the same block hash in an infinite loop, and you have to SIGKILL bitcoind to get things sane again. 

    - HASH BLOK -
0000000000000000049a05148cbe116061e07f2a3f21dcc577b98850603c3072
    - HASH BLOK -
0000000000000000049a05148cbe116061e07f2a3f21dcc577b98850603c3072
    - HASH BLOK -
0000000000000000049a05148cbe116061e07f2a3f21dcc577b98850603c3072
    - HASH BLOK -
    ..

D9B4BEF9 commented:
Alright, some quick progress into what but not why.  

`./src/bitcoind -printtoconsole -zmqpubhashblock=tcp://127.0.0.1:28332'`

The node comes up, that socket listens as expected, and then the node catches up on the latest block. The worker gets stuck emitting the same block hash in an infinite loop, and you have to SIGKILL bitcoind to get things sane again. 

    - HASH BLOK -
0000000000000000049a05148cbe116061e07f2a3f21dcc577b98850603c3072
    - HASH BLOK -
0000000000000000049a05148cbe116061e07f2a3f21dcc577b98850603c3072
    - HASH BLOK -
0000000000000000049a05148cbe116061e07f2a3f21dcc577b98850603c3072
    - HASH BLOK -
    ...

luke-jr commented:
Please cherry-pick and squash 5b2845feabe01f5b671c87bdb909b5b05202d068 and 52396aa65b9e2da7f82e5fd12d87b7a614493ca0

Then you just need to make sure copyright headers are on all new files.

promag commented:
@luke-jr done, thanks!

D9B4BEF9 commented:
@jonasschnelli 

A clean build environment seems to have sorted out my problem of ZMQ just pegging a core.

Diapolo commented on src/init.cpp:
```diff
@@ -31,7 +31,9 @@
 #include "wallet/walletdb.h"
 #endif
-
+#if ENABLE_ZMQ
```
Nit: IMHO the ZMQ header stuff should go below std headers and below boost, openssl etc (header group ordering).

Diapolo commented on src/init.cpp:
```diff
@@ -191,6 +197,17 @@ void Shutdown()
 #endif
+
+#if ENABLE_ZMQ
+    if (pzmqNotificationInterface)
```
Nit: ``if (pzmqNotificationInterface) {`` as sometime in the future this will be the default code formatting (clang cleanup).

Diapolo commented on src/init.cpp:
```diff
@@ -354,6 +371,14 @@ std::string HelpMessage(HelpMessageMode mode)
 #endif
 
+#if ENABLE_ZMQ
+    strUsage += HelpMessageGroup(_("ZeroMQ notification options"));
```
Nit: Misses a ``:`` at the end.

Diapolo commented on src/zmq/zmqconfig.h:
```diff
@@ -0,0 +1,30 @@
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef BITCOIN_ZMQGLOBAL_H
```
This needs to be ``BITCOIN_ZMQ_ZMQCONFIG_H`` also don't forget the header end comment.

Diapolo commented on src/zmq/zmqconfig.h:
```diff
@@ -0,0 +1,30 @@
+
+void zmqError(const char *str);
+
+#if defined(HAVE_CONFIG_H)
```
This is duplicated and also the includes look weird, perhaps some merge-conflict-leftover?

Diapolo commented on src/zmq/zmqabstractnotifier.h:
```diff
@@ -0,0 +1,43 @@
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef BITCOIN_ZMQABSTRACTNOTIFIER_H
```
Need to be ``BITCOIN_ZMQ_ZMQABSTRACTNOTIFIER_H`` also as header end comment.

Diapolo commented on src/zmq/zmqnotificationinterface.cpp:
```diff
@@ -0,0 +1,159 @@
+#include "util.h"
+#include "netbase.h"
+
+#include <zmq.h>
```
Is this needed, seems it's coming from the custom zmq headers.

Diapolo commented on src/zmq/zmqpublishnotifier.cpp:
```diff
@@ -0,0 +1,170 @@
+    }
+    else
+    {
+        LogPrint("zmq", "  Reuse socket for address %s\n", address.c_str());
```
Unneeded ``.c_str()`` and there are some more in here.

Diapolo commented on src/zmq/zmqpublishnotifier.h:
```diff
@@ -0,0 +1,42 @@
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef BITCOIN_ZMQPUBLISHNOTIFIER_H
```
Needs to be ``BITCOIN_ZMQ_ZMQPUBLISHNOTIFIER_H`` also at the header end comment.

promag commented:
@diapolo thanks will fix it later.

jgarzik commented:
"it works" lightly tested ACK.  Needs additional testing to make sure it builds/works on a few platforms.

promag commented:
@venzen Please update and try again.

venzen commented:
@promag  thanks, now compiles and my SUB receives blocks and transactions. However, after an initial one minute spurt of PUBs, Core apparently stops sending `getdata` requests (peer connection activity continues), and an RPC `getrawmempool` shows the mempool is now empty. `bitcoind` process doesn't TERM and has to be killed with -KILL .

Still trying to figure out why. Happens with mainnet, testnet, ipv4 and onion. Tried libzmq1, libzmq3. Compiled with libboost-1.55.0 in Ubuntu 14.04 on two separate hosts.

venzen commented:
@promag OK, I've found the reason why I was getting the behavior mentioned above. It has to do with which notifiers one specifies on the commandline, and also on the combination of notifiers. Some combinations work, other don't. There are similar errors for different combinations.

Firstly, these combinations work as expected:
* hashblock + hashtx
* rawblock + rawtx 
* rawblock + hashtx 
* hashblock + rawtx 
 
Specifying **one notifier only**, has the following results:
* **hashtx**    - inventory exchange stops after *Connect postprocess* and *Connect block* shows in debug.log, p2p connections, `version` and `verack` continues but no `getdata` requests are made.
* **rawtx**     - same as above
* **hashblock** - node starts, creates the PUB socket but then infintely (and rapidly) repeats *Publish hash block*.
* **rawblock**  - same as above but more slowly

Combinations:
* **rawblock + rawtx + hashtx**    - inventory exchange stops after *Connect postprocess* and *Connect block* shows in debug.log, p2p connections, `version` and `verack` continues.
* **rawblock + rawtx + hashblock** - works until hashblock is PUBd and then infinitely repeats *Publish hash block*
* **hashblock + hashtx + rawtx**    - same as above
* **hashblock + hashtx + rawblock** - works until a hashblock is PUBd, then infintely PUBs both hashblock and rawblock, e.g.:

*2015-07-30 05:04:03 Publish hash block 000000000000000014ea2a49bc2165bbc9351adeb9e26870d8ba9049325fe506
2015-07-30 05:04:03 Publish raw block 000000000000000014ea2a49bc2165bbc9351adeb9e26870d8ba9049325fe506*
(repeat infinitely)

Multiple identical notifier instances (this is interesting but probably irrelevant):
* hashblock + 2x hashtx - works
* rawtx + 2x hashblock  - works
* rawbloc + 2x hashtx   - hashtxs are PUBd, rawblock socket is created but no rawblocks PUBd
* hashblock + 2x rawtx  - works

promag commented:
@venzen Thanks for the detailed report. Will look into it.

promag commented:
@venzen can you specify the zmq addresses for each case? Were you using the same address for multiple topics? tpc or ipc?

promag commented on src/zmq/zmqnotificationinterface.cpp:
```diff
@@ -0,0 +1,157 @@
+
+void CZMQNotificationInterface::UpdatedBlockTip(const uint256 &hash)
+{
+    for (std::list<CZMQAbstractNotifier*>::iterator i = notifiers.begin(); i!=notifiers.end(); ++i)
```
Found bug here: 3rd statement `++i` shouldn't be there.
@venzen I believe this is related to you problem, because with only one publisher the loop doesn't ends.

promag commented:
@venzen Fixed and rebased. Tested your combinations and it's working. Please test again.

promag commented:
@jonasschnelli @jgarzik @luke-jr @theuni @sipa do you want to review/test?

jtimon commented:
Untested re-ACK. I'm looking forward to this being finally merged.

jonasschnelli commented on configure.ac:
```diff
@@ -137,6 +137,12 @@ AC_ARG_ENABLE([glibc-back-compat],
   [use_glibc_compat=no])
 
+AC_ARG_ENABLE([zmq],
+  [AC_HELP_STRING([--disable-zmq],
```
I think this should be `--enable-zqm` (@theuni?)

jonasschnelli replied:
nit: still think we we should use `--enable-zmq` with a default `yes`. This would conform with `--enable-wallet`, etc.

jgarzik replied:
Agree with the nit (though it's not a merge blocker)

promag commented:
By default if libzmq is found then it will use unless it's disabled. That was discussed.

promag commented:
@jonasschnelli by default if libzmq is found then it will enabled zmq unless it's disabled on configure. That was discussed: 
https://github.com/bitcoin/bitcoin/pull/4594#issuecomment-50409819.

jonasschnelli commented on src/main.cpp:
```diff
@@ -2303,6 +2303,8 @@ bool ActivateBestChain(CValidationState &state, const CBlock *pblock) {
                         pnode->PushInventory(CInv(MSG_BLOCK, hashNewTip));
             }
             // Notify external listeners about the new tip.
+
```
nit: Remove this line?

jonasschnelli commented on src/zmq/zmqnotificationinterface.cpp:
```diff
@@ -0,0 +1,157 @@
+        CZMQAbstractNotifier *notifier = *i;
+        if (notifier->NotifyBlock(hash))
+        {
+            i ++;
```
nit: s/i ++/i++?

jonasschnelli commented on contrib/zmq/zmq_sub.py:
```diff
@@ -0,0 +1,37 @@
+        body = msg[1]
+
+        if topic == "hashblock":
+            print "- HASH BLOK -"
```
nit: HASH BLOK  (missing C)

promag commented:
@jonasschnelli fixed and rebased.

jonasschnelli commented:
Did some testing.

RPC Test is broken, please cherry pick: https://github.com/jonasschnelli/bitcoin/commit/4e1859a0a55291e67bb8acf3028039668ecfd32b

RPC test sometimes fail (not always).
When calling `generate(10)` sometimes it will result in:
(11 transactions but only 9 blocks)

```
hashtx
hashtx
hashtx
hashblock
hashtx
hashblock
hashtx
hashblock
hashtx
hashblock
hashtx
hashblock
hashtx
hashblock
hashtx
hashblock
hashtx
hashblock
hashtx
hashblock
```

dcousens commented:
concept ACK

promag commented:
@jonasschnelli thanks! However the problem is in the test, see https://github.com/jonasschnelli/bitcoin/commit/4e1859a0a55291e67bb8acf3028039668ecfd32b#commitcomment-12948902.
Fixed your commit and eventually will squash. Please check again.

jonasschnelli commented on qa/pull-tester/rpc-tests.sh:
```diff
@@ -32,6 +32,7 @@ testScripts=(
     'reindex.py'
     'decodescript.py'
     'p2p-fullblocktest.py'
+    'zmq_test.py'
```
I think we need to move the `zmq_test.py` to the `testScriptsExt` list otherwise travis cannot find the zmq python module. Maybe there is also a way to make travis support python/zmq.

promag replied:
I'll remove for now, waiting @theuni feedback.

promag commented:
FYI how I run the test: `python qa/rpc-tests/zmq_test.py --srcdir src/`

jonasschnelli commented:
@promag: Yes. Thats right. The `rpc-tests.sh` will get called during travis CI builds.

theuni commented:
Travis will support it as long as it's installed via depends. However Windows isn't using it right now.

We'll need to add a variable to qa/pull-tester/tests-config.sh.in similar to the ENABLE_WALLET case. 

There are a few other issues I noticed in the configure.ac changes, I'll fix those things up and provide a patch.

promag commented:
@theuni just bumping in case you forget.

theuni commented:
@promag Thanks, indeed I did. I'll get to this asap.

theuni commented:
@promag it's going to be a few days before I can get to the configure changes, but the rpc-tests should be pretty trivial:
```patch
diff --git a/qa/pull-tester/tests-config.sh.in b/qa/pull-tester/tests-config.sh.in
index 10f4d33..e881a95 100755
--- a/qa/pull-tester/tests-config.sh.in
+++ b/qa/pull-tester/tests-config.sh.in
@@ -10,6 +10,7 @@ EXEEXT="@EXEEXT@"
 @ENABLE_WALLET_TRUE@ENABLE_WALLET=1
 @BUILD_BITCOIN_UTILS_TRUE@ENABLE_UTILS=1
 @BUILD_BITCOIND_TRUE@ENABLE_BITCOIND=1
+@ENABLE_ZMQ_TRUE@ENABLE_ZMQ=1
 
 REAL_BITCOIND="$BUILDDIR/src/bitcoind${EXEEXT}"
 REAL_BITCOINCLI="$BUILDDIR/src/bitcoin-cli${EXEEXT}"
```
Then in qa/pull-tester/rpc-tests.sh, ENABLE_ZMQ will either be empty, or set to "1" depending on whether it was enabled in configure. You can use that to conditionally add zmq_test.py to the tests array.

jtimon commented:
Needs rebase (I'm asuming after #5677 ).

promag commented:
@theuni WDYT?

theuni commented:
@promag looks great, assuming it works as intended

promag commented:
@laanwj @sipa any chance this feature be in the next release?

sipa commented on contrib/zmq/zmq_sub.py:
```diff
@@ -0,0 +1,37 @@
+            print '- HASH TX -'
+            print binascii.hexlify(body)
+        elif topic == "rawblock":
+            print "- RAW BLOK HEADER -"
```
s/BLOK/BLOCK/

sipa commented on src/zmq/zmqpublishnotifier.cpp:
```diff
@@ -0,0 +1,170 @@
+
+bool CZMQPublishRawBlockNotifier::NotifyBlock(const uint256 &hash)
+{
+    LogPrint("zmq", "Publish raw block %s\n", hash.GetHex());
```
You need a cs_main lock here.

sipa replied:
Alternatively, if you send CBlockIndex* pointers in the signal rather than the hash, you don't need to. It's the access to mapBlockIndex that's problematic.

jonasschnelli replied:
If I'm right and understand @sipa idea correct, `ReadBlockFromDisk()` not requires a `cs_main` lock. Passing a `CBlockIndex*` would make `CBlockIndex* pblockindex = mapBlockIndex[hash];` at L153 not needed.
The whole cs_main LOCK could then be removed.

We should definitively do this to keep bitcoin-core more responsive and lock free.

sipa commented:
I'm not convinced that the current implementation of sending a full block ("rawblock") makes sense.

The hashes published by UpdatedBlockTip indicate a new block tip, which in the case of a reorganization from chain A-B to A-C-D, will only notify for B and for D, but not for C. This is useful - it's an indication that you need to go query block D, and if you don't have its ancestor, query that too etc.

But when sending full blocks through the interface, I'm much less certain this is useful. You can't be guaranteed to receive all blocks - only those that happen to be at the tip after every reorg/connect.

sipa commented on src/zmq/zmqpublishnotifier.h:
```diff
@@ -0,0 +1,42 @@
+// Copyright (c) 2009-2010 Satoshi Nakamoto
```
This does not seem to contain any code written by Satoshi, or before 2014 :)

promag replied:
Replace with just `// Copyright (c) 2015 The Bitcoin Core developers`?

sipa replied:
Or your own name.

sipa commented on src/validationinterface.cpp:
```diff
@@ -43,6 +45,8 @@ void UnregisterAllValidationInterfaces() {
     g_signals.SetBestChain.disconnect_all_slots();
     g_signals.UpdatedTransaction.disconnect_all_slots();
     g_signals.SyncTransaction.disconnect_all_slots();
+    g_signals.UpdatedTransaction.disconnect_all_slots();
```
You duplicated UpdatedTransaction here.

sipa commented on doc/zmq.md:
```diff
@@ -0,0 +1,116 @@
+The ZeroMQ facility implements a notification interface through a
+set of specific notifiers. Currently there are notifiers that publish
+blocks and transactions. The  supports a set of no binds a "publish"
+port that broadcasts all newly arrived *and validated* blocks and
```
This is not actually correct in the current implementation. You'd need to use CValidationInterface if you want that (see an example in rpcmining.cpp for submitblock).

I think it's very important to make this distinction: what does the ZMQ block interface notify you about? Validated blocks or the accepted block chain? The first can include things like orphaned blocks, and does not tell you which block chain we're currently considering active. The second (what is currently implemented) only notifies for changes in the tip, and may skip blocks in reorgs.

promag replied:
Will review this.

promag commented:
@sipa 

> I'm not convinced that the current implementation of sending a full block ("rawblock") makes sense.

I'm aware of that.

> I think it's very important to make this distinction: what does the ZMQ block interface notify you about? Validated blocks or the accepted block chain?

Publishing blocks happens the exact same way `blocknotify` script is executed. 

The way we plan to use this feature is to replace notify scripts and use as a probe. The subscriber needs to determine the changes since the last notification. If he chooses to receive full block or not it will depend of his use case, after all, it's an option.

promag commented:
@sipa 

> Code looks good, but as mentioned elsewhere, I don't think this is useful
in its current form. If the receiver only receives the blocks that happen
to be at the tip after a reorg, they need a separate mechanism to fetch the
potential unannounced ancestors anyway, making the ability to receive full
blocks immediately unnecessary and confusing.

The subscriber must always query from his last state, because the publisher doesn't know the subscriber state. Suppose the subscriber is restarted, there is some system maintenance or a network issue, the publisher has no way to re-publish what the subscriber needs.

sipa commented:
I'm just questioning what relaying full blocks adds (as opposed to just
hashes), because the receiver should still have logic to fetch missing
blocks in between the previous tip and the new tip.

Again: if you switch from a branch A-B-C-D for example to A-B-C'-D'-E, the
current code will only publish E, and not C' or D'. The publisher still
needs to fetch E's ancestors one by one with the current code. If the
receiver is supposed to do that anyway, what does giving the full block
directly add? Either give all blocks along the path, or just publish the
top hash.

jtimon commented:
Those 2 things are unrelated, full blocks are only allowed optionally to maintain a consistent interface with the transactions (and other types that may be supported un the future). It may be the case that nobody subscribes to raw blocks (and I believe @promag didn't wanted to support full blocks). Take into account that we may add push/pull and/or req/rep in the future (let's finally do this first though).

I don't understand your point about the publisher having to fetch, as promag has explained the publisher cannot know the state of its potentially many subscribers (with potentially different states). Subscribers will need to complement with the RPC (at least until we have req/rep zmq) to get whatever else they need.

promag commented:
+1 what @jtimon said.

sipa commented:
I disagree. What you're publishing is not blocks. It is updates to the chain tip. Sure, there is an actual block associated with that tip, but it's not a natural mapping. You could as well give the chainwork of that tip, or the time.

The documentation is equally confusing: "As transactions and blocks arrive via the P2P network, Bitcoin Core will apply all the validation and standardization rules to this data, only passing along through ZeroMQ those items that pass.". That is not what this pull request implements. It *will not* relay every block through ZeroMQ. It will only ever relay blocks that happen to at the tip of the chain after a reorg.

I think there are a few ways to improve this:
* Only publish the chain tip updates as hashes. That's exactly what -blocknotify does. The subscriber is expected to fetch the actual blocks whenever a notification happens.
* After a reorg, publish *all blocks* that were activated through ZMQ. That will allow a subscriber who is already up-to-date to be able to follow new blocks as they come in and stay in sync. If you want to, I'll implement this.
* Make the block data subscriber trigger BlockChecked and not UpdatedTip. This given the subscriber no information about what the currently active chain is, but it does give all valid blocks as soon as they are validated.

jtimon commented:
> Only publish the chain tip updates as hashes. That's exactly what -blocknotify does. The subscriber is expected to fetch the actual blocks whenever a notification happens.

I believe this is the expected usage rigt now (even if it full blocks don't make much sense with this).


> After a reorg, publish all blocks that were activated through ZMQ. That will allow a subscriber who is already up-to-date to be able to follow new blocks as they come in and stay in sync, without needing to query RPC or REST. If you want to, I'll implement this.

> Make the block data subscriber trigger BlockChecked and not UpdatedTip. This gives the subscriber no information about what the currently active chain is, but it does give all valid blocks as soon as they are validated.

I'm not opposed to the other 2 solutions, I think I like the second one the most.
But if it's going to delay this. Can we at least merge a subset of this with the transactions part?  That would at least introduce the zmq building dependency once and for all? I've been waiting for that for quite some time now...

promag commented:
Sure, the doc needs fixing. The option `-blocknotify` states
```
 -blocknotify=<cmd>
       Execute command when the best block changes (%s in cmd is replaced by
       block hash)
```
And I would go for something like that.

2nd is not viable, since the publisher should not assume the subscriber is up to date. The subscriber should have the option to not receive the (all supposed) missing blocks and he should have the option to defer the fetch of missing blocks. I'm fine with 3rd.

Having the full block on the notification can be helpful: a dummy subscriber, without RPC credential, could just log timestamp, blocksize, tx count for analytics. 

Like @jtimon said, the ideal solution is to have REQ/REP.

sipa commented:
I think that supporting full blocks encourages people to use it in an
incorrect way, by only using those blocks and not taking reorganizations
into account.

promag commented:
How does just the hash encourages to take reorganizations into account?

sipa commented:
Just the hash is perfectly fine. That's what blocknotify does.

Sending the full block corresponding to the new tip encourages bad
practice. People may not realize that it does not relay all blocks.

promag commented:
I still don't see how just the hash solves that problem. People will get the hash, query the block and then it's the same as receiving the full block.

sipa commented:
No, they need to get the hash, query the block, see if they have the
ancestor of the block, if not, query that, etc.

If you give them the blocks already, it will look like they don't need this
recursive fetch logic for ancestors. And it will work fine, except when
there is a reorg, in which they only receive the new tip, and not the block
before.

jgarzik commented:
In general, agree w/ @sipa, just the hash on the first iteration.

Longer term, it is forseeable that some folks looking for low latency may prefer full blocks versus added round-trips.  Such solutions would have a fast path - zero round trip - and a slower path fallback in case the full block was insufficient, and further block-locator style queries are needed.

promag commented:
So it's the same: a client gets the full block, check if the parent is there, if not retrieve and repeat.

promag commented:
@jgarzik also agree. I just don't agree that having just the hash option leads the subscriber to take reorgs into account.

jtimon commented:
Yes, I fail to see the difference as well.

sipa commented:
@promag If implemented correctly, there is no difference. You can present it as an optimization: "This informs you about the new chain tip, and as courtesy, we already give you the last block immediately so you don't need to fetch it". But I think the current interface gives the impression that it will relay all validated blocks, which is wrong and leads to incorrect understanding. The documentation even says so literally.

sipa commented on doc/zmq.md:
```diff
@@ -0,0 +1,115 @@
+connection of a corresponding ZeroMQ subscriber port in receiving 
+software; it is not authenticated nor is there any two-way protocol
+involvement. Therefore, subscribers should validate the received data
+since it may be out of date, incomplete or even invalid.
```
The subscriber has no way to verify that the published block is invalid - it would need a full node for that, and that's exactly what the ZMQ interface is for.

promag replied:
Yeah, reviewing and updating the doc.

sipa replied:
Thanks!

jtimon commented:
Then only the documentation needs to be fixed?
On Sep 10, 2015 6:02 PM, "Pieter Wuille" <notifications@github.com> wrote:

> @promag <https://github.com/promag> If implemented correctly, there is no
> difference. You can present it as an optimization: "This informs you about
> the new chain tip, and as courtesy, we already give you the last block
> immediately so you don't need to fetch it". But I think the current
> interface gives the impression that it will relay all validated blocks,
> which is wrong and leads to incorrect understanding. The documentation even
> says so literally.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/bitcoin/bitcoin/pull/6103#issuecomment-139394307>.
>

sipa commented:
I think many people don't grasp the difference between relaying blocks and chain tip updates, and the fact that this pull request was written with documentation that confuses the two is an indication of that. Don't make the simplest way of using it the wrong one.

sipa commented:
Anyway, it seems my opinion is controversial here. I'm fine with merging if the documentation is fixed, but I would really prefer a mechanism that doesn't make it this easy to misunderstand.

sipa commented:
Also, sorry for hammering on this one detail. I'm very happy with the pull request overall and would like to see it merged soon.

promag commented:
My point is that even `-blocknotify` leads to that problem, and it should be noted elsewhere that a reorg may not call the script for some blocks.

jtimon commented:
> Anyway, it seems my opinion is controversial here. I'm fine with merging
if the documentation is fixed, but I would really prefer a mechanism that
doesn't make it this easy to misunderstand.

Assuming users read the updated docs, I don't see how offering noth
only-hash or full blocks is more confusing than only offering hash-only.

sipa commented:
@jtimon Assuming users read the updated docs...

@promag Agree, it could be clearer there too.

promag commented:
@jonasschnelli @jgarzik @jtimon please review the documentation change.

jonasschnelli commented on src/zmq/zmqpublishnotifier.cpp:
```diff
@@ -0,0 +1,176 @@
+// Copyright (c) 2009-2010 Satoshi Nakamoto
+// Copyright (c) 2009-2014 The Bitcoin Core developers
```
nit: remove `L1` AND `s/2009-2014/2015`?

jonasschnelli commented:
Just git a build error on debian:
```
./configure --enable-debug --with-gui=no
[...]
checking whether to build ZMQ support... yes
checking for ZMQ... yes
[...]
make -j5
[...]
  CXX      wallet/libbitcoin_wallet_a-walletdb.o
zmq/zmqpublishnotifier.cpp: In function ‘int zmq_send_multipart(void*, const void*, size_t, ...)’:
zmq/zmqpublishnotifier.cpp:36:61: error: ‘zmq_msg_send’ was not declared in this scope
         rc = zmq_msg_send(&msg, sock, data ? ZMQ_SNDMORE : 0);
[...]
zmq/zmqnotificationinterface.cpp:118:33: error: ‘zmq_ctx_destroy’ was not declared in this scope
         zmq_ctx_destroy(pcontext);                                                          ^
```

The source of the problem is probably an outdated libzmq. I have installed libzmq1(-dev):
`i A libzmq1                                                 - lightweight messaging kernel (shared library)`

During configure we might ask for `zmq_msg_send` to detect if the library is available (or similar)?

jonasschnelli commented:
Using `libzmq3` solved the problem. But IIRC, Debian 7 (wheezy) does not support libzmq3 out of the standard repository. A tiny check during `./configure` would be nice.

jonasschnelli commented on src/zmq/zmqnotificationinterface.cpp:
```diff
@@ -0,0 +1,156 @@
+#include "main.h"
+#include "streams.h"
+#include "util.h"
+#include "netbase.h"
```
nit: i think this include can be removed. Compiles fine on OSX/Debian without this include.

jonasschnelli commented on src/zmq/zmqpublishnotifier.cpp:
```diff
@@ -0,0 +1,175 @@
+
+#include "zmqpublishnotifier.h"
+#include "main.h"
+#include "net.h"
```
nit: I think this include can be removed. Compiles fine on OSX/Debian without this include.

jtimon replied:
Are you sure this is not because net.h is included in main.h? The same goes for the other removed include [ie netbase.h is included in net.h iirc].

sipa replied:
Whenever you use something defined by a header H in X.c, H needs to be
#included by either X.c or X.h.

You should not remove #includes because they're already included by one of
the dependencies. That leads to unclear dependency graphs, and may cause
breakage when dependencies change.

promag commented:
> Using libzmq3 solved the problem. But IIRC, Debian 7 (wheezy) does not support libzmq3 out of the standard repository. A tiny check during ./configure would be nice.

@theuni can you take a look?

jgarzik commented:
In general, this appears merge-ready once
* travis is happy
* the @sipa lock comment is addressed.

We can iterate further once it's in-tree.

jonasschnelli commented:
Agree with @jgarzik. There are serval proposal for improvements (autodetect libzmq3, avoid locking by passing `CBlockIndex*`, minor cleanups), but shouldn't hold back the merge.
The only thing that might should be looked at is the travis issue: https://travis-ci.org/bitcoin/bitcoin/jobs/79866093. Maybe its not related to this PR. Maybe a rebase&force-push solves that issue.

promag commented:
@jonasschnelli rebase@force-push solved.
@laanwj @sipa I believe it's all in place to merge.
